python/tskit/text_formats.py

```diff
@@ -0,0 +1,85 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Module responsible for working with text format data.
+"""
+import numpy as np
+
+import tskit
+
+
+def parse_fam(fam_path: str, tc: tskit.TableCollection):
+    """
+    Parse PLINK .fam file and convert to tskit IndividualTable.
+
+    Assumes fam file contains five columns: FID, IID, PAT, MAT, SEX
+
+    :param str fam_path: Path to PLINK .fam file
+    :param tskit.TableCollection tc: TableCollection with IndividualTable to
+        which the individuals will be added
+    """
+    individuals = np.loadtxt(
+        fam_path,
+        dtype=str,
+        ndmin=2,  # read file as 2-D table
+        usecols=(0, 1, 2, 3, 4),  # only keep FID, IID, PAT, MAT, SEX columns
+    )  # requires same number of columns in each row, i.e. not ragged
+
+    if len(individuals) == 0:
+        return tc
+
+    id_map = {}  # dict for translating PLINK ID to tskit IndividualTable ID
+    for tskit_id, (plink_fid, plink_iid, _pat, _mat, _sex) in enumerate(individuals):
+        # need this to guarantee uniqueness of individual ID
+        plink_id = (
+            str(plink_fid) + " " + str(plink_iid)
+        )  # include space between strings to ensure uniqueness
+        if plink_id in id_map:
+            raise ValueError("Duplicate PLINK ID")
+        id_map[plink_id] = tskit_id
+    id_map["0"] = "-1"
+
+    tb = tc.individuals
+    tb.metadata_schema = tskit.MetadataSchema(
+        {
+            "codec": "json",
+            "type": "object",
+            "properties": {
+                "plink_fid": {"type": "string"},
+                "plink_iid": {"type": "string"},
+                "sex": {"type": "integer"},
+            },
+            "required": ["plink_fid", "plink_iid", "sex"],
+            "additionalProperties": True,
+        }
+    )
+    for plink_fid, plink_iid, pat, mat, sex in individuals:
+        sex = int(sex)
+        if not (sex in range(3)):
+            raise ValueError(
+                "Sex must be one of the following: 0 (unknown), 1 (male), 2 (female)"
+            )
+        metadata_dict = {"plink_fid": plink_fid, "plink_iid": plink_iid, "sex": sex}
+        pat = plink_fid + " " + pat if pat != "0" else pat
+        mat = plink_fid + " " + mat if mat != "0" else mat
+        tb.add_row(parents=[id_map[pat], id_map[mat]], metadata=metadata_dict)
```

Comment:
I think we need to do a topological sort here on the parents table to make sure that it can be used as a valid tree sequence. Doing ``tables.sort()`` should do it.

Reply:
We should test this by making an example where some children come before their parents, and then calling ``ts = tables.tree_sequence()`` on it.